CARDS-2263: Store HERACLES S3 data export task performance metrics in the JCR

IntegralProgrammer opened the pull request:
This _Pull Request_ introduces 3 new performance metrics, all related to HERACLES S3 data exports, namely:

- `S3ExportedForms` counts the number of _Forms_ successfully pushed to the configured S3 bucket
- `S3ExportedSubjects` counts the number of Subjects successfully pushed to the configured S3 bucket
- `S3ExportFailures` counts the number of S3 export tasks that have terminated in error

Additionally, an `ErrorTracking` service is made available so that whenever `ErrorTracking.logError` is called, an `nt:file` node in the JCR under `/LoggedEvents` is created which holds the contents of the logged error. The `SlackNotificationsTask` is modified so that any logged errors under `/LoggedEvents` are sent in the periodic Slack alerts update along with the performance metrics recorded under `/Metrics`. As of the current implementation, all logged errors under `/LoggedEvents` will be sent to Slack, therefore once the errors have been resolved, these node under `/LoggedEvents` should be manually deleted so that alerts of these (now resolved) errors are no longer sent to Slack.

As the errors collected by `ErrorTracking.logError` can potentially be long, I needed to switch from using basic _text_ objects on the Slack API to using _attachment_ objects. This automatically adds a _Show more / Show less_ button for messages that would be too long to be sent to Slack using basic _text_ objects. This, however, slightly changes the appearance of the message of all our Slack alert messages.

For example, a nightly performance metric update message from _Cards4PREMs_ would now look like this,

![screenshot_2023_07_12-15-30-19](https://github.com/data-team-uhn/cards/assets/5349728/a2baba55-f1da-4032-8a05-8dc1f83b1af8)

instead of this,

![screenshot_2023_07_12-15-30-32](https://github.com/data-team-uhn/cards/assets/5349728/5224aa0d-9fd1-49a7-bb48-d1bb8a4d434b)

Testing Instructions
--------------------------

1. Build this (`CARDS-2263`) branch including a Docker image with `mvn clean install -Pdocker`.
2. `cd compose-cluster`
3. `python3 generate_compose_yaml.py --dev_docker_image --oak_filesystem --composum --cards_project cards4heracles --slack_notifications --s3_test_container`.
4. You will be prompted to enter a Slack webhook URL. Either create your own Slack channel and app with an _incoming webhook_ or DM me to be added to the _#cards-alerts-dev_ Slack channel and I will provide you with an incoming webhook.
5. Edit `docker-compose.yml` and add the environment variable `NIGHTLY_SLACK_NOTIFICATIONS_SCHEDULE=0 * * ? * *` to the `cardsinitial` container.
6. `docker-compose build && docker-compose up -d`
7. Every minute, a performance metric update containing **Number Of Forms Exported to S3 bucket**, **Number Of Subjects Exported to S3 bucket**, and **Number Of Failed S3 export jobs** should be sent to Slack.
8. Visit http://localhost:8080 and login as `admin`:`admin`
9. Create a _Demographics_ Form and a test _Patient_ Subject.
10. Visit http://localhost:8080/Subjects.s3push?dateLowerBound={TODAY}&dateUpperBound={TOMORROW}. For example, if today is July 12th 2023, you would visit http://localhost:8080/Subjects.s3push?dateLowerBound=2023-07-12&dateUpperBound=2023-07-13.
11. The next update in Slack should show **Number Of Failed S3 export jobs** incremented by 1 followed by a stack trace showing a `NoSuchBucket` error.
12. Visit http://localhost:8080/bin/browser.html and delete the `nt:file` node under `/LoggedEvents`
13. Visit http://localhost:9001 and login as `minioadmin`:`minioadmin`
14. Click on _Create bucket_ and create a new bucket called `uhn`.
15. Visit the `Subjects.s3push` URL once again (eg. http://localhost:8080/Subjects.s3push?dateLowerBound=2023-07-12&dateUpperBound=2023-07-13)
16. The next Slack update should increment both **Number Of Forms Exported to S3 bucket** and **Number Of Subjects Exported to S3 bucket** by 1.
17. Create a new _Baseline Health Information_ Form for the same test patient and visit the S3 push URL (eg. http://localhost:8080/Subjects.s3push?dateLowerBound=2023-07-12&dateUpperBound=2023-07-13) once again.
18. The next Slack update should increment **Number Of Forms Exported to S3 bucket** by 2 and **Number Of Subjects Exported to S3 bucket** by 1.
19. Visit http://localhost:8080/bin/browser.html and delete everything under `/Metrics`.
20. The next Slack update should show _**ERROR:** Could not gather any performance statistics_.
21. `docker-compose down && docker-compose rm && docker volume prune -f && ./cleanup.sh`

IntegralProgrammer committed:
CARDS-2263: Store HERACLES S3 data export task performance metrics in the JCR

Create the /Metrics JCR node and its children

IntegralProgrammer committed:
CARDS-2263: Store HERACLES S3 data export performance metrics in the JCR

Store stack traces of io.uhndata.cards.heracles.internal.export failures in the JCR

IntegralProgrammer committed:
CARDS-2263: Store HERACLES S3 data export performance metrics in the JCR

Increment the S3ExportFailures metric whenever io.uhndata.cards.heracles.internal.export throws an exception

IntegralProgrammer committed:
CARDS-2263: Store HERACLES S3 data export performance metrics in the JCR

Increment the S3ExportedSubjects metric whenever io.uhndata.cards.heracles.internal.export successfully exports a Subject

IntegralProgrammer committed:
CARDS-2263: Store HERACLES S3 data export performance metrics in the JCR

Increment the S3ExportedForms metric whenever io.uhndata.cards.heracles.internal.export successfully exports a Form

IntegralProgrammer committed:
CARDS-2263: Store HERACLES S3 data export performance metrics in the JCR

The cards-metrics service user needs jcr:read access to /LoggedEvents/

IntegralProgrammer committed:
CARDS-2263: Store HERACLES S3 data export performance metrics in the JCR

Display any logged events (stack traces) with each Slack update message

IntegralProgrammer committed:
CARDS-2263: Store HERACLES S3 data export performance metrics in the JCR

Fix bug where "Could not gather any performance statistics" would be incorrectly displayed in green

sashaandjic reviewed (approved):
tested

acrowthe reviewed (approved):
Reviewed and tested, looks good

sdumitriu commented on distribution/src/main/features/base-repoinit.txt:
```diff
@@ -32,6 +32,16 @@ set principal ACL for cards-metrics
+# error-tracking
+create path (sling:Folder) /LoggedEvents
+
+create service user cards-exporter with path system/sling
+set principal ACL for cards-exporter
+    allow   jcr:read    on /
+    allow   jcr:addChildNodes,jcr:modifyProperties,jcr:nodeTypeManagement   on /LoggedEvents
 end
```
The error tracking feature should be self contained, meaning that it should use its own service user, since it could be used by other modules, not just the "exporter". So:
- create a `cards-error-logger` service user here
- add a user mapping in the error-tracker's feature file from the error-tracker module to the new cards-error-logger user

sdumitriu replied:
And I forgot to mention, this should be a core service, so it should always be included in the distribution by adding a new prototype feature file in `distribution/src/main/features/core`.

IntegralProgrammer replied:
>     * add a user mapping in the error-tracker's feature file from the error-tracker module to the new cards-error-logger user

Is that even possible to do? When I try to do so, the error log shows an error `The provided service user id 'serviceuser--io.uhndata.cards.heracles-backend--ErrorTracking' is not a known JCR system user id and therefore not allowed in the Sling Service User Mapper.`. So it looks like `io.uhndata.cards.heracles-backend:ErrorTracking` would need to be mapped to `cards-error-logger` as opposed to `io.uhndata.cards.errortracking` being mapped to `cards-error-logger`. Furthermore, I see that in other places in our code base, for example, in https://github.com/data-team-uhn/cards/blob/ba6d2bc72ef344e67df3fdc3c359d2c759faca1c/proms-resources/feature/src/main/features/feature.json#L68-L73, we need to specify the mapping between `io.uhndata.cards.proms-backend:MetricLogger` and `cards-metrics` as opposed to specifying a mapping between `io.uhndata.cards.metrics:MetricLogger` and `cards-metrics`.

I like the idea of only needing to specify the mapping once in the `feature.json` file for the `error-tracking` module but would that even be possible? Do we have any examples in our code base where we do that?

sdumitriu replied:
I think the issue is that you're using the resource resolver factory obtained for the wrong module. Once you inject one directly in the error logger module, the correct module name should be in use. At least in theory.

IntegralProgrammer replied:
Yes, that's correct, and I've just implemented that with - https://github.com/data-team-uhn/cards/pull/1563/commits/cd87854a18596933ceb4ef189244b263d4e60c9e, so it's now a _service_ as opposed to just a library of functions. I'll refactor the code a bit to address your other comments but as you can that fixes the `io.uhndata.cards.errortracking` vs `io.uhndata.cards.heracles` issue.

sdumitriu commented on distribution/src/main/features/base-repoinit.txt:
```diff
@@ -32,6 +32,16 @@ set principal ACL for cards-metrics
+# error-tracking
+create path (sling:Folder) /LoggedEvents
+
+create service user cards-exporter with path system/sling
```
Why `with path`?

sdumitriu commented on distribution/src/main/features/base-repoinit.txt:
```diff
@@ -32,6 +32,16 @@ set principal ACL for cards-metrics
+    allow    jcr:read    on /LoggedEvents/
+end
+
+# error-tracking
```
Shouldn't all this be done in `modules/error-tracking/src/main/features/feature.json`? Or rather, in `feature-repoinit.txt`.

IntegralProgrammer replied:
Implemented in https://github.com/data-team-uhn/cards/pull/1563/commits/a6acc30700b262fa6f5d6f30c643f5a5b79b903e :heavy_check_mark:.

sdumitriu commented on heracles-resources/backend/src/main/java/io/uhndata/cards/heracles/internal/export/ExportTask.java:
```diff
@@ -99,6 +103,15 @@ public void run()
+            StringWriter sw = new StringWriter();
+            PrintWriter pw = new PrintWriter(sw);
+            e.printStackTrace(pw);
+            ErrorTracking.logError(this.resolverFactory, sw.toString());
```
I would add a `logError(ResolverFactory, Throwable)` method that does the printing inside. Or even better, without the `ResolverFactory`.

IntegralProgrammer replied:
Implemented in https://github.com/data-team-uhn/cards/pull/1563/commits/a05e674b5d4e6c5cc70536570c154e140cf0e6ad :heavy_check_mark:

sdumitriu commented on heracles-resources/backend/src/main/java/io/uhndata/cards/heracles/internal/export/ExportTask.java:
```diff
@@ -99,6 +103,15 @@ public void run()
         } catch (Exception e) {
             LOGGER.error("Failed to perform the nightly export", e.getMessage(), e);
+
+            // Create an nt:file node under /LoggedEvents/ storing the stack trace of this failure
```
This is too technical.

```suggestion
            // Store the stack trace of this failure
```

IntegralProgrammer replied:
Fixed - https://github.com/data-team-uhn/cards/pull/1563/commits/69db72ea0a9afa9884193b478c76453ba238d80a :heavy_check_mark:

sdumitriu commented on modules/error-tracking/src/main/java/io/uhndata/cards/errortracking/ErrorTracking.java:
```diff
@@ -0,0 +1,83 @@
+     * @param stackTrace the String containing the stack trace of the error that was thrown resulting
+     * in the calling of this method
+     */
+    public static void logError(final ResourceResolverFactory resolverFactory, final String stackTrace)
```
I don't like that we have to pass the resolver factory here. I also think having this as a service too would be useful. And I also think that `ErrorLogger` is a better name. So, I think this would work:

- add `ErrorLogger` as an interface with two methods, `log(String stackTrace)` and `log(Throwable throwable)`
- add `StaticErrorLogger` as a static class with the same methods, plus a `setService(ErrorLogger service)` method
- add `ErrorLoggerImpl` as a service implementation, with a reference to `ResourceResolverFactory`
- make it an immediate service with `@Component(immediate=true)`
- add an `@Activate` method that calls `StaticErrorLogger.setService(this)

WDYT?

IntegralProgrammer replied:
Implemented in https://github.com/data-team-uhn/cards/pull/1563/commits/daeae99ef631a646ebad69aa866b3240c8360a6a :heavy_check_mark:

sdumitriu commented on modules/slack-notifications/src/main/java/io/uhndata/cards/slacknotifications/SlackNotificationsTask.java:
```diff
@@ -115,6 +145,9 @@ public void run()
             }
 
+            // Get all the error stack traces under /LoggedEvents/
+            List<String> errorStackTraces = getLoggedEvents(resolver);
```
Won't the notifications get longer and longer with each new error logged? I think we shouldn't dump all the actual logs, but either:
- the count of total errors
- the count of new errors, which would involve also storing the error date, maybe as the `jcr:created` of each file node
- the root cause class name + counts of new errors, which would involve storing the root cause class name with each error

I think the last one would be ideal, but not urgent for now. Something like:

```
The following new errors were logged:
- 2 * NullPointerException
- 4 * RepositoryException
```

IntegralProgrammer replied:
> Won't the notifications get longer and longer with each new error logged?

Yes, that is true, like I have discussed in the beginning of this Pull Request.

> As of the current implementation, all logged errors under `/LoggedEvents` will be sent to Slack, therefore once the errors  have been resolved, these nodes under `/LoggedEvents` should be manually deleted so that alerts of these (now resolved)  errors are no longer sent to Slack.

So, for now, the purpose of CARDS-2263, is simply to alert administrators via Slack if something is going wrong. Once the issue is resolved, the logged error event nodes can be deleted so that they are no longer sending notifications to Slack. Perhaps a nice improvement to this would be to add a JCR node which keeps track of the latest manual fix and only send to Slack logged errors that were `jcr:created` _after_ the date/time specified by the latest manual fix node. That is pretty much what we are doing in https://github.com/data-team-uhn/cards-log-monitor/blob/master/generate_alerts/generate_alerts.py which we are using for monitoring the logs for the Cards4CaRe project.

WDYT?

sdumitriu replied:
OK, sounds good.

sdumitriu commented on modules/slack-notifications/src/main/java/io/uhndata/cards/slacknotifications/SlackNotificationsTask.java:
```diff
@@ -128,10 +161,18 @@ public void run()
-                slackNotificationString = "*ERROR*: Could not gather any performance statistics";
+            // Include any relevant stack traces
+            Iterator<String> stackTracesIter = errorStackTraces.iterator();
+            while (stackTracesIter.hasNext()) {
```
I think a separator message at this point would be useful, e.g. "The following errors were logged:".

IntegralProgrammer replied:
Addressed in https://github.com/data-team-uhn/cards/pull/1563/commits/e5a497c4d54afd6da11eaee16a4dac9ba83c6f09.

![screenshot_2023_08_15-17-16-39](https://github.com/data-team-uhn/cards/assets/5349728/6e8f542e-8aef-4905-b61d-d319c25bd400)

IntegralProgrammer committed:
CARDS-2263: Store HERACLES S3 data export performance metrics in the JCR

If there are stack traces of errors to be displayed, add a separator message - "The following errors were logged:"

IntegralProgrammer committed:
CARDS-2263: Store HERACLES S3 data export performance metrics in the JCR

Fix a comment

IntegralProgrammer committed:
CARDS-2263: Store HERACLES S3 data export performance metrics in the JCR

Obtain the ResourceResolverFactory in the ErrorTracking module so that JCR access permissions can be set for it as opposed to needing to set its permissions from a calling module

sashaandjic commented:
@IntegralProgrammer  Is this PR now ready for re-test?

sdumitriu commented:
> @IntegralProgrammer Is this PR now ready for re-test?

Not yet.

IntegralProgrammer committed:
CARDS-2263: Store HERACLES S3 data export performance metrics in the JCR

Code cleanup: No longer need to pass a ResourceResolverFactory object to the ErrorTracking.logError method

IntegralProgrammer committed:
CARDS-2263: Store HERACLES S3 data export performance metrics in the JCR

Generation of the error stack trace string is now done in the ErrorTracking.logError method

IntegralProgrammer committed:
CARDS-2263: Store HERACLES S3 data export performance metrics in the JCR

Moved the creation of the /LoggedEvents jcr:Folder and the error-tracking service user into the error-tracking feature-repoinit.txt file

IntegralProgrammer committed:
CARDS-2263: Store HERACLES S3 data export performance metrics in the JCR

Moved the repository setup of the cards-metrics module into its own feature-repoinit.txt file and made the cards-metrics module a core CARDS module

IntegralProgrammer committed:
CARDS-2263: Store HERACLES S3 data export performance metrics in the JCR

Refactor ErrorTracking.java into ErrorLogger.java, ErrorLoggerImpl.java, and StaticErrorLogger.java

sdumitriu commented on modules/error-tracking/src/main/features/feature-repoinit.txt:
```diff
@@ -0,0 +1,27 @@
+# error-tracking
+create path (sling:Folder) /LoggedEvents
+
+create service user error-tracking with path system/sling
```
```suggestion
create service user error-tracking
```

IntegralProgrammer replied:
Without the `with path system/sling` part, nothing is written into the JCR under `/LoggedEvents/`.

sdumitriu commented on distribution/src/main/features/base-repoinit.txt:
```diff
@@ -24,14 +24,9 @@ set principal ACL for sling-readall
-    allow    jcr:write   on /Metrics/
-    allow    jcr:addChildNodes    on /Metrics/
-    allow    jcr:nodeTypeManagement    on /Metrics/
+create service user cards-exporter with path system/sling
+set principal ACL for cards-exporter
+    allow   jcr:read    on /
```
These should be moved in the feature file for heracles.

sdumitriu commented on distribution/src/main/features/base-repoinit.txt:
```diff
@@ -24,14 +24,9 @@ set principal ACL for sling-readall
-    allow    jcr:write   on /Metrics/
-    allow    jcr:addChildNodes    on /Metrics/
-    allow    jcr:nodeTypeManagement    on /Metrics/
+create service user cards-exporter with path system/sling
```
`with path` is not needed, and it's not right to use `system/sling` as the target path, since this is not a Sling user.

```suggestion
create service user cards-exporter
```

sdumitriu commented on heracles-resources/backend/src/main/java/io/uhndata/cards/heracles/internal/export/ExportTask.java:
```diff
@@ -99,6 +101,12 @@ public void run()
+            StaticErrorLogger.logError(e);
+
+            // Increment the count of S3ExportFailures
+            Metrics.increment(this.resolverFactory, "S3ExportFailures", 1);
```
Looks like the Metrics component could use the same improvement the ErrorLogger had, not having to pass the resolver factory for every call. But that should be done in a separate PR.

sdumitriu commented on modules/error-tracking/src/main/java/io/uhndata/cards/errortracking/ErrorLoggerImpl.java:
```diff
@@ -0,0 +1,101 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@Component(immediate = true, service = ErrorLoggerImpl.class)
```
Why not:

```suggestion
@Component(immediate = true, service = ErrorLogger.class)
```

sdumitriu commented on modules/error-tracking/src/main/java/io/uhndata/cards/errortracking/ErrorLoggerImpl.java:
```diff
@@ -0,0 +1,101 @@
+    @Activate
+    protected void activate(ComponentContext componentContext) throws Exception
+    {
+        LOGGER.warn("ErrorTracking IS ACTIVATING!");
```
Remove for final version.

```suggestion
```

sdumitriu commented on modules/error-tracking/src/main/java/io/uhndata/cards/errortracking/ErrorLoggerImpl.java:
```diff
@@ -0,0 +1,101 @@
+        StaticErrorLogger.setService(this);
+    }
+
+    /*
+     * Stores in the JCR, under /LoggedEvents, a nt:file node containing the passed stack trace.
+     *
+     * @param loggedError the Throwable containing the stack trace of the error that was thrown resulting
+     * in the calling of this method
+     */
```
```suggestion
    @Override
```

sdumitriu commented on modules/error-tracking/src/main/java/io/uhndata/cards/errortracking/ErrorLoggerImpl.java:
```diff
@@ -0,0 +1,101 @@
+            final String newFileName = UUID.randomUUID().toString();
+            final Map<String, Object> eventNodeProperties = new HashMap<>();
+            eventNodeProperties.put("jcr:primaryType", "nt:file");
+            resolver.create(eventsFolderResource, newFileName, eventNodeProperties);
+            Resource thisEventResource = resolver.getResource(LOGGED_EVENTS_PATH + newFileName);
```
```suggestion
            Resource thisEventResource = resolver.create(eventsFolderResource, newFileName, eventNodeProperties);
```

sdumitriu commented on modules/error-tracking/src/main/java/io/uhndata/cards/errortracking/StaticErrorLogger.java:
```diff
@@ -0,0 +1,52 @@
+    {
+    }
+
+    public static void setService(ErrorLogger service)
```
```suggestion
    static void setService(ErrorLogger service)
```

sdumitriu commented on modules/error-tracking/src/main/java/io/uhndata/cards/errortracking/StaticErrorLogger.java:
```diff
@@ -0,0 +1,52 @@
+     */
+    public static void logError(final Throwable loggedError)
+    {
+        errorLoggerService.logError(loggedError);
```
Should check for `null` first:

```suggestion
        if (errorLoggerService != null) {
            errorLoggerService.logError(loggedError);
        }
```

sdumitriu commented on modules/metrics/src/main/features/feature-repoinit.txt:
```diff
@@ -0,0 +1,31 @@
+create service user cards-metrics with path system/sling
+
+set principal ACL for cards-metrics
+    allow    jcr:read    on /Metrics/
+    allow    jcr:write   on /Metrics/
+    allow    jcr:addChildNodes    on /Metrics/
+    allow    jcr:nodeTypeManagement    on /Metrics/
```
```suggestion
    allow    jcr:read,jcr:write,jcr:addChildNodes,jcr:nodeTypeManagement    on /Metrics/
```

sdumitriu commented on modules/metrics/src/main/features/feature-repoinit.txt:
```diff
@@ -0,0 +1,31 @@
+# cards-metrics
+create path (sling:Folder) /Metrics
+
+create service user cards-metrics with path system/sling
```
```suggestion
create service user cards-metrics
```

IntegralProgrammer replied:
This change breaks periodic sending of messages to Slack - on Slack, the error `ERROR: Could not gather any performance statistics` is shown instead of the metrics listed in the JCR under `/Metrics`.

sdumitriu replied:
That doesn't sound like something removing `with path`  would cause.

sdumitriu commented on modules/patient-portal/src/main/features/feature.json:
```diff
@@ -37,10 +37,6 @@
-    {
-      "id":"${project.groupId}:cards-metrics:${project.version}",
-      "start-order":"25"
-    },
```
Why this removal? Patient Portal still uses metrics.

IntegralProgrammer replied:
Yes, but given that I have made the `metrics` module a core module in https://github.com/data-team-uhn/cards/pull/1563/commits/ab1eb1a6de10bfdbaabe6322c4b5189ffd258ef4, does it still need to be specified here? I tested in _Cards4PREMs_ mode and can confirm that metrics still increment as expected when a user submits a survey.

sdumitriu replied:
It doesn't need to be specified, but so do most of the other dependencies. Listing them here ensures that the order in which different modules are loaded is correct, so basically what we want is the `start-order` to match.

sdumitriu reviewed (changes requested):
I think I'd like a small rename:
- ErrorLogger -> ErrorLoggerService
- StaticErrorLogger -> ErrorLogger

sdumitriu committed:
CARDS-2263: Store HERACLES S3 data export performance metrics in the JCR
Move heracles repoinit statements in the heracles feature file

sdumitriu committed:
CARDS-2263: Store HERACLES S3 data export performance metrics in the JCR
Don't put cards system users under /system/sling

sdumitriu committed:
CARDS-2263: Store HERACLES S3 data export performance metrics in the JCR
Rename ErrorLogger to ErrorLoggerService and StaticErrorLogger to ErrorLogger

sdumitriu committed:
[cleanup] Remove debug code

sdumitriu committed:
CARDS-2263: Store HERACLES S3 data export performance metrics in the JCR
Avoid NPE

sdumitriu committed:
CARDS-2263: Store HERACLES S3 data export performance metrics in the JCR
Don't duplicate inherited javadoc

sdumitriu committed:
CARDS-2263: Store HERACLES S3 data export performance metrics in the JCR
Simplify Sling API usage

sdumitriu committed:
CARDS-2263: Store HERACLES S3 data export performance metrics in the JCR
Reduce method visibility

sdumitriu committed:
CARDS-2263: Store HERACLES S3 data export performance metrics in the JCR
Simplify repoinit statements

IntegralProgrammer commented on heracles-resources/feature/src/main/features/feature.json:
```diff
@@ -83,10 +83,33 @@
+        "set properties on /Metrics/S3ExportedSubjects/prevTotal \n   default value{Long} to 0 \n end",
+        "set properties on /Metrics/S3ExportFailures/name \n   default value{String} to \"{003} Number Of Failed S3 export jobs\" \n end",
+        "set properties on /Metrics/S3ExportFailures/prevTotal \n   default value{Long} to 0 \n end",
+        "create service user cards-exporter \n set principal ACL for cards-exporter \n   allow jcr:read on / \n end"
```
```suggestion
        "create service user cards-exporter with path system/sling \n set principal ACL for cards-exporter \n   allow jcr:read on / \n end"
```

Without specifying the `with path system/sling` part, S3 export jobs do not run.

IntegralProgrammer commented:
Tested 3 times with the `with path system/sling` code _present_ as was originally in this PR and for all 3 times, expected data was sent to Slack.

Tested 3 times with the `with path system/sling` code _removed_ as was introduced in the changes made by @sdumitriu and for all 3 times, the text `ERROR: Could not gather any performance statistics` was sent to Slack.

sdumitriu commented:
I figured out why it doesn't work.

`set principal ACL`  only works for principals under `/users/system/sling`, since the principal-based ACLs are not quite standard, but an Oak extension to the default ACL defined by the JCR specification, which isn't enabled by default. Sling enables it, and configures it to only be valid for sling system users.

Either use `... with path` when creating the system user, use `set acl` instead of `set principal acl` to set rights, or configure the principal ACL implementation to support all system users.

I'll fix it.

sdumitriu committed:
CARDS-2263: Store HERACLES S3 data export task performance metrics in the JCR
Make principal-based ACLs work for all system users, not just sling ones
Move a few of our system users under system/cards

sdumitriu commented on heracles-resources/feature/src/main/features/feature.json:
```diff
@@ -83,10 +83,33 @@
       "user.mapping":[
-        "io.uhndata.cards.heracles-backend=[sling-readall]",
-        "io.uhndata.cards.heracles-backend:PauseResumeEditor=[sling-readall]"
+        "io.uhndata.cards.heracles-backend=[cards-exporter]",
```
Make this a subservice as well:
```suggestion
        "io.uhndata.cards.heracles-backend:S3Export=[cards-exporter]",
```

Requires changes in Java as well, when obtaining the resource resolver.

sdumitriu committed:
CARDS-2263: Store HERACLES S3 data export task performance metrics in the JCR
Define a subservice for the S3 export

sdumitriu reviewed (approved):
I think this is ready to merge.

IntegralProgrammer commented on heracles-resources/feature/src/main/features/feature.json:
```diff
@@ -83,10 +83,33 @@
+        "set properties on /Metrics/S3ExportFailures/prevTotal \n   default value{Long} to 0 \n end",
+        "create service user cards-exporter with path system/cards \n set principal ACL for cards-exporter \n   allow jcr:read on / \n end"
+      ]
+    },
     "org.apache.sling.serviceusermapping.impl.ServiceUserMapperImpl.amended~heracles":{
       "user.mapping":[
-        "io.uhndata.cards.heracles-backend=[sling-readall]",
-        "io.uhndata.cards.heracles-backend:PauseResumeEditor=[sling-readall]"
+        "io.uhndata.cards.heracles-backend:S3Export=[cards-exporter]",
```
I think `cards-exporter` sounds too generic to be placed in the project specific `heracles-resources` module. Maybe it should be renamed to `heracles-exporter`?

WDYT @sdumitriu?

sdumitriu replied:
Since it has read privileges on `/`, we could just use `sling-readall`.

sashaandjic commented:
I am not getting this error:
![image](https://github.com/data-team-uhn/cards/assets/23243244/89e14bf2-e2bf-448c-b7d3-b4201cd8e556)

If you look at slack the error is:
![image](https://github.com/data-team-uhn/cards/assets/23243244/6d9c1814-c1a6-4534-9a46-dc853f1c40cd)

Also after creating S3 bucket, there is no export happening.

The only thing that worked as expected is :
![image](https://github.com/data-team-uhn/cards/assets/23243244/db5fca99-e32c-443b-b269-02f9f0aa6082)

sdumitriu committed:
CARDS-2263: Store HERACLES S3 data export task performance metrics in the JCR
More efficient use of resource resolvers

sashaandjic reviewed (approved):
tested
![image](https://github.com/data-team-uhn/cards/assets/23243244/6a0e9ab5-4070-4f2d-91fb-caca87ad2a67)
